sycl/doc/Assert.md

```diff
@@ -0,0 +1,426 @@
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+Whenever `isAssertEnabled` property set is present, this specific device image
+was built with `NDEBUG` macro undefined and it requires fallback implementation
+of `__devicelib_assert_fail` (i.e. if Device-side Runtime doesn't support it).
+
+Any properties in `isAssertEnabled` property set are ignored.
+
+The property set is added to device binary descriptor whenever at least single
+translation unit was compiled with assertions enabled i.e. `NDEBUG` undefined.
+
+The property set is added by `sycl-post-link` tool depending on module metadata.
+Metadata is provided by Clang frontend. Metadata name is `is_assert_enabled`.
+
+
+##### Compiling with assert enabled/disabled
+
+Consider the following example sources:
+```c++
+// impl.cpp
+using namespace sycl;
+int calculus(int X) {
+  assert(X && "Invalid value");
+  return X * 2;
+}
+
+void enqueueKernel(queue &Q, buffer &B) {
+  Q.submit([](handler &H) {
+    auto Acc = B.get_access<read_write>(H);
+    H.parallel_for(/* range */, [](item It) {
+      assert(Acc[It]);
+      // ...
+    });
+  });
+}
+
+// main.cpp
+// ...
+using namespace sycl;
+
+SYCL_EXTERNAL int calculus(int);
+void enqueueKernel(queue&, buffer&);
+
+void workload() {
+  queue Q;
+  buffer B;
+
+  Q.submit([](handler &H) {
+    auto Acc = B.get_access<read_write>(H);
+    H.parallel_for(/* range */, [](item It) {
+      int X = calculus(0); // should fail assertion
+      assert(X && "Nil in result");
+      Acc[It] = X;
+    });
+  });
+
+  enqueueKernel(Q, B);
+  ...
+}
+```
+
+These two files are compiled into a single binary application. There are four
+states of definedness of `NDEBUG` macro available:
+
+| # | `impl.cpp` | `main.cpp` |
+| - | ---------- | ---------- |
+| 1 | defined    | defined    |
+| 2 | defined    | undefined  |
+| 3 | undefined  | defined    |
+| 4 | undefined  | undefined  |
+
+States of definedness of `NDEBUG` macro defines the set of assertions which can
+fail. Having assertions enabled in at least one translation unit with device
+code requires for `isAssertEnabled` property set being present in device image
+descriptor structure.
+
+
+### Storing accessor metadata and writing assert failure to buffer
+
+Both storing of accessor metadata and writing assert failure is performed with
+help of built-ins. Implementations of these builtins are substituted by
+frontend.
+
+User's kernel is executed through a wrapper. Wrapping takes place in DPCPP
+Runtime headers in a following manner:
+
+```
+class handler {
+
+template <typename KernelName> parallel_for(KernelFunc, Range) {
+#ifndef NDEBUG
+  // Assert required
+  if (!MQueue->get_device()->assert_fail_supported()) {
+    using KName2 = class ASSERT_WRAPPER_NAME(KernelName);
+    
+    auto AssertBufferAcc = MQueue->get_context()->getAssertBufferAccessor(this);
```

Comment:
If I understand correctly, program scope variables in OpenCL are only persistent within a single OpenCL program, which means that this buffer should be created per device image: in case when device code split is enabled, each kernel could go into a separate device image, i.e. into a separate OpenCL program.

Reply:
Even though, the device images are distinct the allocation is the same within single context. This allows one to disregard splitness of device binaries.

The program-scope variable should be emitted for each and every device image which contains a kernel calling `__store_acc`/`__store_assert_info` built-ins.

Reply:
> Even though, the device images are distinct the allocation is the same within single context. This allows one to disregard splitness of device binaries.

Oh, yes, mistaken program scope variable with the buffer. Please disregard my comment